source/sigv4.c

```diff
@@ -1162,110 +1620,1287 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+
+    if( returnStatus == SigV4Success )
+    {
+        returnStatus = validateDateTime( &date );
+    }
+
+    if( returnStatus == SigV4Success )
+    {
+        /* Combine date elements into complete ASCII representation, and fill
+         * buffer with result. */
+        intToAscii( date.tm_year, &pWriteLoc, ISO_YEAR_LEN );
+        intToAscii( date.tm_mon, &pWriteLoc, ISO_NON_YEAR_LEN );
+        intToAscii( date.tm_mday, &pWriteLoc, ISO_NON_YEAR_LEN );
+        *pWriteLoc = 'T';
+        pWriteLoc++;
+        intToAscii( date.tm_hour, &pWriteLoc, ISO_NON_YEAR_LEN );
+        intToAscii( date.tm_min, &pWriteLoc, ISO_NON_YEAR_LEN );
+        intToAscii( date.tm_sec, &pWriteLoc, ISO_NON_YEAR_LEN );
+        *pWriteLoc = 'Z';
+
+        LogDebug( ( "Successfully formatted ISO 8601 date: \"%.*s\"",
+                    ( int ) dateISO8601Len,
+                    pDateISO8601 ) );
+    }
+
+    return returnStatus;
+}
+
+SigV4Status_t SigV4_GenerateHTTPAuthorization( const SigV4Parameters_t * pParams,
+                                               char * pAuthBuf,
+                                               size_t * authBufLen,
+                                               char ** pSignature,
+                                               size_t * signatureLen )
+{
+    SigV4Status_t returnStatus = SigV4Success;
+    CanonicalContext_t canonicalContext;
+    const char * pAlgorithm = NULL;
+    char * pSignedHeaders = NULL;
+    size_t encodedLen = 0U, algorithmLen = 0U, signedHeadersLen = 0U, authPrefixLen = 0U;
+    HmacContext_t hmacContext = { 0 };
+    SigV4String_t signingKey;
+
+    returnStatus = verifySigV4Parameters( pParams );
+
+    authPrefixLen = *authBufLen;
+
+    /* Default arguments. */
+    if( ( pParams->pAlgorithm == NULL ) || ( pParams->algorithmLen == 0 ) )
+    {
+        /* The default algorithm is AWS4-HMAC-SHA256. */
+        pAlgorithm = SIGV4_AWS4_HMAC_SHA256;
+        algorithmLen = SIGV4_AWS4_HMAC_SHA256_LENGTH;
+    }
+    else
+    {
+        pAlgorithm = pParams->pAlgorithm;
+        algorithmLen = pParams->algorithmLen;
+    }
+
+    if( returnStatus == SigV4Success )
+    {
+        returnStatus = generateCanonicalRequestUntilHeaders( pParams, &canonicalContext, &pSignedHeaders, &signedHeadersLen );
+    }
+
+    /* Write the prefix of the Authorizaton header value. */
+    if( returnStatus == SigV4Success )
+    {
+        returnStatus = generateAuthorizationValuePrefix( pParams,
+                                                         pAlgorithm, algorithmLen,
+                                                         pSignedHeaders, signedHeadersLen,
+                                                         pAuthBuf, &authPrefixLen );
+    }
+
+    /* Hash and hex-encode the canonical request to the buffer. */
+    if( returnStatus == SigV4Success )
+    {
+        encodedLen = canonicalContext.bufRemaining;
+        returnStatus = completeHashAndHexEncode( pParams->pHttpParameters->pPayload,
```

Comment:
Hashing the payload again may not be required as Application is already hashing the payload and passing it as value of `x-amz-content-sha56` header. We can directly use that value. I have discussed it with @aggarw13 already.

Reply:
I have created a [separate branch in my fork](https://github.com/aggarw13/SigV4-for-AWS-IoT-embedded-sdk/tree/hygiene/optimize-payload-hashing) for logic updates of the optimization. I will create a separate PR for it.

Reply:
@gshvang @aggarw13 Should we add another option instead that allows the customer to choose whether to hash the payload or not?